ntoskrnl/ex/locale.c

```diff
@@ -24,8 +24,250 @@ LANGID PsInstallUILanguageId = LANGIDFROMLCID(0x00000409);
+ *   in the Alternate Sorts key;
+ *
+ * - The locale must exist in the Language Groups key, and
+ *   the queried value must be readable;
+ *
+ * - The locale registry data must be of REG_SIZE type,
+ *   has a consistent length and the locale is set.
+ *
+ * @param[in] LocaleId
+ * A locale identifier that corresponds to a specific
+ * locale to be validated.
+ *
+ * @return
+ * Returns STATUS_SUCCESS if the function has successfully
+ * validated the locale and it is valid. STATUS_OBJECT_NAME_NOT_FOUND
+ * is returned if the following locale does not exist on the system.
+ * A failure NTSTATUS code is returned otherwise.
+ */
+static
+NTSTATUS
+ExpValidateNlsLocaleId(
+    _In_ LCID LocaleId)
+{
+    NTSTATUS Status;
+    HANDLE NlsLocaleKey = NULL, AltSortKey = NULL, LangGroupKey = NULL;
+    OBJECT_ATTRIBUTES NlsLocalKeyAttrs, AltSortKeyAttrs, LangGroupKeyAttrs;
+    PKEY_VALUE_PARTIAL_INFORMATION BufferKey;
+    WCHAR ValueBuffer[20], LocalBuffer[20], ValueIdBuffer[20];
+    USHORT BufferLength;
+    PWCHAR Id;
+    ULONG i, IdLength, ReturnedLength;
+    UNICODE_STRING LocaleName;
+    UNICODE_STRING LocaleIdString;
+    static UNICODE_STRING NlsLocaleKeyPath = RTL_CONSTANT_STRING(L"\\Registry\\Machine\\SYSTEM\\CurrentControlSet\\Control\\Nls\\Locale");
+    static UNICODE_STRING AltSortKeyPath = RTL_CONSTANT_STRING(L"\\Registry\\Machine\\SYSTEM\\CurrentControlSet\\Control\\Nls\\Locale\\Alternate Sorts");
+    static UNICODE_STRING LangGroupPath = RTL_CONSTANT_STRING(L"\\Registry\\Machine\\SYSTEM\\CurrentControlSet\\Control\\Nls\\Language Groups");
+
+    /* Initialize the registry path attributes */
+    InitializeObjectAttributes(&NlsLocalKeyAttrs,
+                               &NlsLocaleKeyPath,
+                               OBJ_CASE_INSENSITIVE | OBJ_KERNEL_HANDLE,
+                               NULL,
+                               NULL);
+
+    InitializeObjectAttributes(&AltSortKeyAttrs,
+                               &AltSortKeyPath,
+                               OBJ_CASE_INSENSITIVE | OBJ_KERNEL_HANDLE,
+                               NULL,
+                               NULL);
+
+    InitializeObjectAttributes(&LangGroupKeyAttrs,
+                               &LangGroupPath,
+                               OBJ_CASE_INSENSITIVE | OBJ_KERNEL_HANDLE,
+                               NULL,
+                               NULL);
+
+    /* Compute the buffer length of the locale name */
+    BufferLength = sizeof(LocalBuffer) / sizeof(WCHAR);
+
+    /* Build the locale name */
+    LocaleName.Length = BufferLength;
+    LocaleName.MaximumLength = BufferLength;
+    LocaleName.Buffer = LocalBuffer;
+
+    /* Convert the locale ID into a suitable string */
+    Status = RtlIntegerToUnicodeString(LocaleId, 16, &LocaleName);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("Failed to convert LCID to a string (Status 0x%lx)\n", Status);
+        return Status;
+    }
+
+    /* Fill the formatted buffer with zeroes */
```

Comment:
Why is this necessary to fill it with _**characters**_ zero ? And not just zeroing-out the buffer "as usual" (just byte zero).

Reply:
I realized the way this comment is phrased is misleading, my bad. That portion of code is used to format the LCID string into a hexadecimal format which we will then use it to check it against the existing locale identifiers which are in a hex format too.